Refactor the networking client

AndrejMitrovic opened the pull request:
It also fixes a small bug where it used vibe.d for sleeping the task instead of using the task manager.

I am doing this to simplify the request retry / banning code I'll be introducing.

AndrejMitrovic committed:
Move RemoteNode to the network package

This is code that deals with networking, request retries,
and error handling, it doesn't contain any business
logic of the node.

AndrejMitrovic committed:
Remove unused getPublicConfig() API

It was implemented too eagerly, the concept
of validator nodes will be introduced at a
later point.

AndrejMitrovic committed:
Get rid of callbacks and rename getReady()

The use of callbacks here didn't really make sense,
as the function blocked the same task that called
the getReady().

Additionally, the retrieval of the client's peers
depends on minPeersConnected(), which is a property
of the NetworkManager - so the code belongs there.

The getReady() function was renamed to handshake().

Currently it's blocking (no change in semantics),
later it might throw an exception if a number of
failed attempts were made to handshake with the node.

AndrejMitrovic committed:
Refactor getNetworkInfo and remove net_info field

The field was only used as a temporary,
we can use a parameter instead.

AndrejMitrovic committed:
Use TaskManager for sleeping a task

This was a bug, it was using vibe.d's sleeping
routines instead of whichever one was provided
by the TaskManager (e.g. LocalRest's).

AndrejMitrovic commented:
Update: fixed private => public being in the wrong commit.

Geod24 commented on source/agora/common/NetworkClient.d:
```diff
@@ -105,15 +105,15 @@ class NetworkClient
             logInfo("[%s] Couldn't retrieve public key. Will retry in %s..",
                 this.address, this.retry_delay);
-            sleep(this.retry_delay);
+            this.taskman.wait(this.retry_delay);
```
Can you remove the `vibe.core.core` import ?

Geod24 commented on source/agora/common/NetworkClient.d:
```diff
@@ -152,18 +151,18 @@ class NetworkClient
     ***************************************************************************/
 
-    private bool getNetworkInfo ()
+    private bool getNetworkInfo (out NetworkInfo net_info)
```
Need to add it to DDOC

Geod24 reviewed (changes requested):
As discussed, please move the client to a `network` package instead of `common`.
Otherwise looks very very good to me.

AndrejMitrovic committed:
Move Network module into the network package

Network-related code should all live under
the same package, separate from the business
logic of the node.

AndrejMitrovic committed:
Rename Network => NetworkManager

This matches the convention in all other modules,
and makes it easier to differentiate between
NetworkManager and the upcoming NetworkClient.

AndrejMitrovic committed:
Rename RemoteNode => NetworkClient

AndrejMitrovic commented:
Updated with suggested changes.

AndrejMitrovic commented:
Green & eco-friendly.
